fix no sent mailbox configured

hamza221 opened the pull request:
Fix #4153 
Fixes https://github.com/nextcloud/mail/issues/10546

Repro: 

1. remove the default folder for sent ("you can do it in vuex by setting sentMailboxId to null")
2. Try to send a message 

Fix: 
Ps: Other ideas are appreciated 
1. Check if a mailbox named Sent or equivalent in the local language

- if it exists set it as Sent default folder and resend message, if not:

2. create a new mailbox 
3. set it as send default folder
4. resend the message 
5. Catch exceptions and inform the user with toasts accordingly

hamza221 committed:
fix: no sent mailbox configured

Signed-off-by: hamza221 <hamzamahjoubi221@gmail.com>

ChristophWurst commented on src/components/Composer.vue:
```diff
@@ -1121,6 +1121,13 @@ export default {
 			this.saveDraftDebounced()
 		},
 		async onSend(_, force = false) {
+			const selectedAccount = this.$store.getters.getAccount(this.selectedAlias.id)
```
does this work with aliases too? I think `getAccount` will only return an object if the id references an account (the main address), not an alias

hamza221 replied:
https://github.com/nextcloud/mail/blob/87c3ac8db07a365b358f71e8a644b584a68289d4/src/components/Composer.vue#L654-L658
selectedAlias.id here is always the account id

ChristophWurst commented on src/components/NewMessageModal.vue:
```diff
@@ -128,6 +135,69 @@ export default {
+			this.creatingSentMailbox = true
+
+			// check if sent mailbox already exists
+			let sentMailboxId = await getMailboxExists(data.accountId, account.personalNamespace + 'Sent').catch((error) => {
```
would it not be sufficient to search for the mailbox in the local data?

JohannesGGE commented:
Where exactly do you set sentMailboxId to null? (I can't get the error reproduced)

hamza221 commented:
>Where exactly do you set sentMailboxId to null? (I can't get the error reproduced)
I do it in the vue.js devtools extension

ChristophWurst commented on src/components/NewMessageModal.vue:
```diff
@@ -128,6 +134,60 @@ export default {
+			logger.info(`creating ${t('mail', 'Sent')} mailbox`)
+			await this.$store
+				.dispatch('createMailbox', { account, name: account.personalNamespace + t('mail', 'Sent') })
+				.then((e) => {
```
the method is async. you don't need .then. use await

ChristophWurst reviewed (commented):
The changes feel a bit too complex and messy. I think we should try to find a cleaner, less complicated spot to place the logic. I'll think about an alternative.

ChristophWurst commented:
> The changes feel a bit too complex and messy. I think we should try to find a cleaner, less complicated spot to place the logic. I'll think about an alternative.

How about we just move the logic into the head of NewMessageModal::onSend? We have access to the accountId and can fetch the account, check for the mailbox existence and act, then proceed with the usual send logic.

The good news is that this isn't too far off the current state. The code just has to be pushed around a bit.

I also think we don't necessarily need a special loading text for this state. For the user it's sufficient if we show *Sending …* while creating/assigning the mailbox

ChristophWurst commented on src/components/NewMessageModal.vue:
```diff
@@ -294,6 +294,51 @@ export default {
+			showWarning(t('mail', 'Setting Sent default folder'))
+			let newSentMailboxId = null
+			const mailboxes = this.mainStore.getMailboxes(data.accountId)
+			const sentMailboxId = mailboxes.find((mailbox) => mailbox.name === account.personalNamespace + 'Sent' || mailbox.name === account.personalNamespace + t('mail', 'Sent'))?.databaseId
```
nit. first look for the localized, then search for `Sent`. this ensures that the localized takes priority.

ChristophWurst commented on src/components/NewMessageModal.vue:
```diff
@@ -294,6 +294,51 @@ export default {
+			const mailboxes = this.mainStore.getMailboxes(data.accountId)
+			const sentMailboxId = mailboxes.find((mailbox) => mailbox.name === account.personalNamespace + 'Sent' || mailbox.name === account.personalNamespace + t('mail', 'Sent'))?.databaseId
+			if (sentMailboxId) {
+				await this.setSentMailboxAndResend(account, sentMailboxId, data)
```
I'm missing a return or anything else that leaves this method after setting the existing mailbox

ChristophWurst replied:
> hamza221 marked this conversation as resolved.

Isn't this precisely the reason why I'm running into errors despite being able to send the message?

To have constructive and comprehensive collaboration please be more elaborative. If you close my comment as resolved silently I assume it's because you have addressed it.

ChristophWurst commented on src/components/NewMessageModal.vue:
```diff
@@ -294,6 +294,51 @@ export default {
+			}
+			logger.info(`creating ${t('mail', 'Sent')} mailbox`)
+			try {
+				const newSentMailbox = await this.mainStore.createMailbox({ account, name: account.personalNamespace + t('mail', 'Sent') })
```
if the IMAP server supports it (likely) we should also set the specialuse attribute to `\sent`

ChristophWurst reviewed (commented):
The logical flow looks good otherwise

ChristophWurst commented on lib/Controller/MailboxesController.php:
```diff
@@ -246,10 +246,10 @@ public function update() {
 	 */
 	#[TrapError]
-	public function create(int $accountId, string $name): JSONResponse {
+	public function create(int $accountId, string $name, bool $isSentMailbox = false): JSONResponse {
```
This makes the new parameter only usable for specialuse=sent. Change it to allow a generic specialuse attribute. Then we will be able to also create other special mailboxes.

ChristophWurst commented on appinfo/routes.php:
```diff
@@ -155,6 +155,11 @@
 			'verb' => 'PATCH'
 		],
+		[
+			'name' => 'mailboxes#setImapSpecialUseAtrribute',
```
dead route

ChristophWurst reviewed (changes requested):
No test coverage

hamza221 committed:
fixup! fix: no sent mailbox configured

Signed-off-by: Hamza Mahjoubi <hamzamahjoubi221@gmail.com>

hamza221 commented:
Creating a new mailbox is not tested, not sure how to without an account

kesselb commented on lib/IMAP/FolderMapper.php:
```diff
@@ -76,8 +86,17 @@ public function getFolders(Account $account, Horde_Imap_Client_Socket $client,
+			}
+			$opts[] = $attribute;
+		}
+		$client->createMailbox($name, $opts);
```
createMailbox expects an array "opts" with a key "special_use" and there the list of attributes.

ChristophWurst commented on lib/IMAP/FolderMapper.php:
```diff
@@ -76,8 +86,17 @@ public function getFolders(Account $account, Horde_Imap_Client_Socket $client,
+	
+		foreach ($specialUseAttributes as $attribute) {
+			if (!in_array($attribute, self::SUPPORTED_SPECIAL_USE_ATTRIBUTES, true)) {
+				throw new ServiceException('Unsupported special use attribute: ' . $attribute);
```
It's sensible validation. You could pull this up into the HTTP layer and do the check in the controller. Return a BAD REQUEST or UNPROCESSIBLE ENTITY.

hamza221 committed:
fixup! fix: no sent mailbox configured

Signed-off-by: Hamza Mahjoubi <hamzamahjoubi221@gmail.com>

ChristophWurst commented on lib/Controller/MailboxesController.php:
```diff
@@ -248,8 +248,12 @@ public function update() {
+		try {
+			$mailbox = $this->mailManager->createMailbox($account, $name, $specialUseAttributes);
+		} catch (ServiceException $e) {
+			return new JSONResponse([], Http::STATUS_BAD_REQUEST);
```
This will now catch *any* service exception and return a nondescriptive  HTTP response. This would be painful to debug.

Do a simple in_array check comparing the sent special use to the allowed ones. Then carry on with the string and add type hints for the allowed values for the deeper service methods.

hamza221 committed:
fixup! fix: no sent mailbox configured

Signed-off-by: Hamza Mahjoubi <hamzamahjoubi221@gmail.com>

hamza221 committed:
fixup! fix: no sent mailbox configured

Signed-off-by: Hamza Mahjoubi <hamzamahjoubi221@gmail.com>

hamza221 committed:
fixup! fix: no sent mailbox configured

Signed-off-by: Hamza Mahjoubi <hamzamahjoubi221@gmail.com>

ChristophWurst committed:
Merge remote-tracking branch 'origin/main' into bug/not-sent-mailbox

ChristophWurst commented:
Works, but not as expected.

![Bildschirmfoto vom 2025-05-16 13-28-28](https://github.com/user-attachments/assets/4ce007cf-91bc-44e3-9a3f-c3c5fdd8fcb7)

I see this chain of toasts, so the logic kicks in.

![Bildschirmfoto vom 2025-05-16 13-29-15](https://github.com/user-attachments/assets/e875e26f-0275-485b-b1bf-a14ff7d7f21d)

But my sent mailbox is now set to *nullSent*. Which is not just a UI glitch, it created this mailbox:

![image](https://github.com/user-attachments/assets/2511e25d-f866-4568-9e6d-b9318e0f68d7)

The creation shouldn't even happen, because I do have a mailbox named **Sent** and it does have the **\\sent** attribute:

![image](https://github.com/user-attachments/assets/33a67593-ac4f-4baa-88b2-45f388ee4f3a)

ChristophWurst commented on src/components/NewMessageModal.vue:
```diff
@@ -286,6 +286,51 @@ export default {
+			}
+			logger.info(`creating ${t('mail', 'Sent')} mailbox`)
+			try {
+				const newSentMailbox = await this.mainStore.createMailbox({ account, name: account.personalNamespace + t('mail', 'Sent'), specialUseAttributes: ['\\Sent'] })
```
I guess the ``nullSent`` happens when there is no personalNamespace. In that case, don't prepend anything

hamza221 committed:
fixup! fix: no sent mailbox configured

Signed-off-by: Hamza <hamzamahjoubi221@gmail.com>

hamza221 committed:
Merge remote-tracking branch 'origin/main' into bug/not-sent-mailbox

hamza221 committed:
fixup! fix: no sent mailbox configured

Signed-off-by: Hamza <hamzamahjoubi221@gmail.com>

hamza221 commented:
> It fails to detect my existing mailbox _Sent_ and runs into a HTTP400. The toasts are honestly overwhelming and confusing.
> 
> First I get this ![Bildschirmfoto vom 2025-05-27 09-52-22](https://private-user-images.githubusercontent.com/1374172/447816891-444ed0f9-656c-4b92-9b7b-df66f636908f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDkxMjIyOTIsIm5iZiI6MTc0OTEyMTk5MiwicGF0aCI6Ii8xMzc0MTcyLzQ0NzgxNjg5MS00NDRlZDBmOS02NTZjLTRiOTItOWI3Yi1kZjY2ZjYzNjkwOGYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDYwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA2MDVUMTExMzEyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDZiYWE2MjlhZTc5NjFjYmNlMGNiMGM1ZTkyNmRkMzAzNThlYzVlNjllNmVjODg4OGJhNmEwYzk0NjQ2ZjdiYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.IPiZGe8bLmIrhwXyXofwbL0KQFzsnt21oaeJXoPWlV4)
> 
> ![Bildschirmfoto vom 2025-05-27 09-52-25](https://private-user-images.githubusercontent.com/1374172/447816961-1213a112-9772-42a6-a65a-3f38ff250055.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDkxMjIyOTIsIm5iZiI6MTc0OTEyMTk5MiwicGF0aCI6Ii8xMzc0MTcyLzQ0NzgxNjk2MS0xMjEzYTExMi05NzcyLTQyYTYtYTY1YS0zZjM4ZmYyNTAwNTUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDYwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA2MDVUMTExMzEyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjUxMzMyYTBlZTAxOGM2ZTEzYTU3ZDMwM2ZkYWUzMGNjZDM5ZmUyYzk3MmY4ODMwNzUzM2YwNGJhY2FmZDNjNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.zkVNzEvVBrtVpvJPxx5361Os6wcoDIVvKBdKnBWaMcc)
> 
> ![image](https://private-user-images.githubusercontent.com/1374172/447817271-02106622-900c-45d0-9440-0f5043a67199.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDkxMjIyOTIsIm5iZiI6MTc0OTEyMTk5MiwicGF0aCI6Ii8xMzc0MTcyLzQ0NzgxNzI3MS0wMjEwNjYyMi05MDBjLTQ1ZDAtOTQ0MC0wZjUwNDNhNjcxOTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDYwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA2MDVUMTExMzEyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTI5ZTI5ODcyY2ZiNThiNWQxYTM0YTdiODY4NDkyODczNTBjZjk2YzBlY2JiMWFkMzBhN2U2Y2Q4YTEyYjBkNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.NS65BaxHe2D00JtIbdnYuunITiY-vzjsMjJZgORzlNQ)
> 
> The message _is sent and stored in my mailbox called "Sent"_, nevertheless?!


Toasts were all over the place, 
It should make sense now, I added early returns and removed extra Toasts

ChristophWurst commented on lib/Contracts/IMailManager.php:
```diff
@@ -45,12 +45,13 @@ public function getMailboxes(Account $account): array;
 	/**
 	 * @param Account $account
 	 * @param string $name
+	 * @param array<Horde_Imap_Client::SPECIAL_USE_*> $specialUseAttributes
```
^ see Psalm. Missing an import

ChristophWurst commented on src/components/NewMessageModal.vue:
```diff
@@ -286,6 +286,57 @@ export default {
+				return
+
+			}
+			logger.info(`creating ${t('mail', 'Sent')} mailbox`)
+			try {
+				const newSentMailbox = await this.mainStore.createMailbox({ account, name: (account.personalNamespace ?? '') + t('mail', 'Sent'), specialUseAttributes: ['\\Sent'] })
```
nit: extract new mailbox name into a variable for code readability and to have the actual mailbox name logged, not just the translation

ChristophWurst replied:
`createMailbox` prepends the namespace too. This can lead to double namespacing.

ChristophWurst commented on src/components/NewMessageModal.vue:
```diff
@@ -286,6 +286,57 @@ export default {
+			const sentMailboxId = mailboxes.find((mailbox) => (mailbox.name === (account.personalNamespace ?? '') + 'Sent') || (mailbox.name === (account.personalNamespace ?? '') + t('mail', 'Sent')))?.databaseId
+			if (sentMailboxId) {
+				try {
+					await this.setSentMailboxAndResend(account, sentMailboxId, data)
```
the method takes to params. `data` is superfluous

ChristophWurst commented:
Superseded by https://github.com/nextcloud/mail/pull/12555
